mapiproxy/libmapistore/mapistore_notification.c

```diff
@@ -3,525 +3,1355 @@
-        if (notification_types == fnevTableModified) {
-                table_parameters = notification_parameters;
-                new_subscription->parameters.table_parameters = *table_parameters;
-        }
-        else {
-                object_parameters = notification_parameters;
-                new_subscription->parameters.object_parameters = *object_parameters;
-
-		/* NewMail POC: open newmail mail queue */
-		if (notification_types & fnevNewMail || notification_types & fnevObjectCreated) {
-			new_subscription->mqueue_name = talloc_asprintf((TALLOC_CTX *)new_subscription, 
-									MAPISTORE_MQUEUE_NEWMAIL_FMT, username);
-			new_subscription->mqueue = mq_open(new_subscription->mqueue_name, 
-							   O_RDONLY|O_NONBLOCK|O_CREAT, 0777, NULL);
-			if (new_subscription->mqueue == -1) {
-				perror("mq_open");
-				talloc_free(new_subscription->mqueue_name);
-				return new_subscription;
-			}
-			
-			/* Empty queue since we only want to retrieve new data from now */
-			ret = mq_getattr(new_subscription->mqueue, &attr);
-			if (ret == -1) {
-				perror("mq_getattr");
-			} else {
-				data.data = talloc_size(mem_ctx, attr.mq_msgsize);
-				while ((data.length = mq_receive(new_subscription->mqueue, (char *)data.data,
-								 attr.mq_msgsize, &prio)) != -1) {
-					dump_data(0, data.data, data.length);
-					talloc_free(data.data);
-					data.data = talloc_size(mem_ctx, attr.mq_msgsize);
-				}
-				talloc_free(data.data);
-			}
-			
-			/* Set destructor on new_subscription as we want to unlink the queue upon release */
-			talloc_set_destructor((void *)new_subscription, (int (*)(void *))mapistore_subscription_destructor);
-			
-			/* Send notification to tell about newmail */
-			n.WholeStore = object_parameters->whole_store;
-			n.NotificationFlags = notification_types;
-			if (n.WholeStore == true) {
-				n.FolderID = 0;
-				n.MessageID = 0;
-				n.MAPIStoreURI = NULL;
-			} else {
-				n.FolderID = object_parameters->folder_id;
-				n.MessageID = object_parameters->object_id;
-				/* FIXME */
-				n.MAPIStoreURI = NULL;
-			}
+	struct mapistore_notification_context	*notification_ctx = NULL;
+	const char				*url = NULL;
+	memcached_server_st			*servers = NULL;
+	memcached_return			rc;
 
-			c.username = (char *)username;
-			c.mstore_ctx = mstore_ctx;
-			
-			ret = mapistore_mgmt_interface_register_subscription(&c, &n);
-			OC_DEBUG(0, ("[%s:%d]: registering notification: %d\n", __FUNCTION__, __LINE__, ret));
-		}
+	/* Sanity checks */
+	MAPISTORE_RETVAL_IF(!lp_ctx, MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+	MAPISTORE_RETVAL_IF(!_notification_ctx, MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+
+	notification_ctx = talloc_zero(mem_ctx, struct mapistore_notification_context);
+	MAPISTORE_RETVAL_IF(!notification_ctx, MAPISTORE_ERR_NO_MEMORY, NULL);
+	notification_ctx->memc_ctx = NULL;
+
+	url = lpcfg_parm_string(lp_ctx, NULL, "mapistore", "notification_cache");
+	if (url) {
+		notification_ctx->memc_ctx = memcached(url, strlen(url));
```

Comment:
The documentation was lacking details and after looking at the implementation, I don't have more confidence about this call. Most of the calling functions are relying on static buffer which length is `BUFSIZ` and there are some use case where error_buffer and  error_buffer_size are not set which would require extra checks.

Too much hassle with regards to expected benefits here.